flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TemporalRowtimeJoin.scala

```diff
@@ -0,0 +1,339 @@
+import org.apache.flink.types.Row
+import org.apache.flink.util.Preconditions.checkState
+
+import scala.collection.JavaConversions._
+
+/**
+  * This operator works by keeping on the state collection of probe and build records to process
+  * on next watermark. The idea is that between watermarks we are collecting those elements
+  * and once we are sure that there will be no updates we emit the correct result and clean up the
+  * state.
+  *
+  * Cleaning up the state drops all of the "old" values from the probe side, where "old" is defined
+  * as older then the current watermark. Build side is also cleaned up in the similar fashion,
+  * however we always keep at least one record - the latest one - even if it's past the last
+  * watermark.
+  *
+  * One more trick is how the emitting results and cleaning up is triggered. It is achieved
+  * by registering timers for the keys. We could register a timer for every probe and build
+  * side element's event time (when watermark exceeds this timer, that's when we are emitting and/or
+  * cleaning up the state). However this would cause huge number of registered timers. For example
+  * with following evenTimes of probe records accumulated: {1, 2, 5, 8, 9}, if we
+  * had received Watermark(10), it would trigger 5 separate timers for the same key. To avoid that
+  * we always keep only one single registered timer for any given key, registered for the minimal
+  * value. Upon triggering it, we process all records with event times older then or equal to
+  * currentWatermark.
+  */
+class TemporalRowtimeJoin(
+    leftType: TypeInformation[Row],
+    rightType: TypeInformation[Row],
+    genJoinFuncName: String,
+    genJoinFuncCode: String,
+    queryConfig: StreamQueryConfig,
+    leftTimeAttribute: Int,
+    rightTimeAttribute: Int)
+  extends AbstractStreamOperator[CRow]
+  with TwoInputStreamOperator[CRow, CRow, CRow]
+  with Triggerable[Any, VoidNamespace]
+  with Compiler[FlatJoinFunction[Row, Row, Row]]
+  with Logging {
+
+  validateEqualsHashCode("join", leftType)
+  validateEqualsHashCode("join", rightType)
+
+  private val NEXT_LEFT_INDEX_STATE_NAME = "next-index"
+  private val LEFT_STATE_NAME = "left"
+  private val RIGHT_STATE_NAME = "right"
+  private val REGISTERED_TIMER_STATE_NAME = "timer"
+  private val TIMERS_STATE_NAME = "timers"
+
+  private val rightRowtimeComparator = new RowtimeComparator(rightTimeAttribute)
+
+  /**
+    * Incremental index generator for `leftState`'s keys.
+    */
+  private var nextLeftIndex: ValueState[JLong] = _
+
+  /**
+    * Mapping from artificial row index (generated by `nextLeftIndex`) into the left side `Row`.
+    * We can not use List to accumulate Rows, because we need efficient deletes of the oldest rows.
+    *
+    * TODO: this could be OrderedMultiMap[Jlong, Row] indexed by row's timestamp, to avoid
+    * full map traversals (if we have lots of rows on the state that exceed `currentWatermark`).
+    */
```

Comment:
From what I get from a quick look at the code, we access elements in the state based on their timestamp. Using a `MapState` with the index as the key, instead of using a plain `ListState` is to avoid having to fetch and serialize/deserialize the whole state. So, if the above is correct and I am not missing something, then why not using a `MapState<Long, List<CRow>>`. This seems to be fulfilling all the above requirements and we avoid multi-state accesses for the same timestamp (assuming that this can happen). 

I will have a closer look at the code later, as I am also working on a relevant issue, but for now this is something that caught my eye.

Reply:
`Map<Long, List<CRow>>` would be significantly less efficient in the more common case with one element per timestamp (in milliseconds). Creating huge number of lists holding a single element would slow down the code and could cause GC issues.

Reply:
As far as the number of lists is concerned, this is not an issue, as we can have a single list that we just `clear()` and repopulate for each element. This I think also partially answers the "significantly slow down for the common usecase" argument (no significant GC oberhead), for which actually I disagree in the first place, as, I would assume, that a RocksDB access is far more expensive than creating a list. In addition, I am not so sure what is the "common" usecase. Under the label "common", I would only put RocksDB as state backend, as this is what we normally recommend for production usecases. What are the other assumptions that we should consider as common? If it is that we have a single element per timestamp, then, again, a single list that we clear and repopulate incurs a minimal overhead.

One final argument is that having a timestamp as a key would improve code readability. The reason why I focused on this in the first place, was that I could not understand the reason why we have as key an arbitrary index, which is an overhead in terms of storage and state accesses.

Reply:
I think the `MapState<Long, CRow>` is a good solution. It features cheap insertions and the iterating the MapState is fairly efficient due to pre-fetching in RocksDB.

Reply:
@kl0u other thing is that I don't fully understand our pressure on using RocksDB, while vast majority of our users (at least that was result of my analysis of `DataStream` usage) would be perfectly fine with Heap state backend. 

Regarding performance penalty, just accessing one more level of collection is a significant slowdown (more memory access), even assuming that we could reuse the Lists - which we couldn't, since keeping the empty list on the state would prevent state cleanup.

Reply:
@pnowojski If everyone else agrees on the current implementation, feel free to go on with this strategy. This was just to express my concerns about the chosen path.

BTW could you share the analysis on state backend usage? I actually have no numbers but just the fact that we recommend RocksDB for production use as it allows you to only be bounded by your disk.